pandas/core/construction.py

```diff
@@ -394,14 +396,14 @@ def extract_array(obj: AnyArrayLike, extract_numpy: bool = False) -> ArrayLike:
     array([1, 2, 3])
     """
     if isinstance(obj, (ABCIndexClass, ABCSeries)):
-        obj = obj.array
+        result = obj.array
+    else:
+        result = obj
 
-    if extract_numpy and isinstance(obj, ABCPandasArray):
-        obj = obj.to_numpy()
+    if extract_numpy and isinstance(result, ABCPandasArray):
+        result = result.to_numpy()
 
-    # error: Incompatible return value type (got "Index", expected "ExtensionArray")
-    # error: Incompatible return value type (got "Series", expected "ExtensionArray")
-    return obj  # type: ignore[return-value]
+    return result
```

Comment:
the problem here is that AnyArrayLike is a typevar and also the type of `obj` is wrong anyway and should be `object` I think could also change `obj` to object/Any for now until we overload and keep the implementation unchanged.

Reply:
can you revert the changes to the implementation.

Reply:
Yes,  but would require adding `obj = obj.array  # typing: ignore[assignment]` or possible changing typing on arguments (not return type)

Reply:
IIUC this is a similar issue to https://github.com/pandas-dev/pandas/pull/37902#discussion_r526021864 where we changed FrameOrSeries to FrameOrSeriesUnion.

We don't yet have an AnyArrayLikeUnion to do the same here.

because AnyArrayLike is a typevar, the type annotations imply that the type doesn't change.

This is not true. Only for np.ndarray and ExtensionArray if extract_array=False.

so maybe the signature should be

```
def extract_array(
    obj: Union[object, ArrayLike], extract_numpy: bool = False
) -> Union[object, ArrayLike]:
```

but we also need to overload (to account for Series and Index array extraction) but while numpy types resolve to `Any` we get mypy errors about overlapping overloads.

so for now, i think ok to do...

```
def extract_array(
    obj: object, extract_numpy: bool = False
) -> Union[Any, ArrayLike]:
```

It's the use of the typevar that causes the issue, so I'm not too concerned what type annotations are chosen at this stage, so long as we are not changing the implementation because we are using a TypeVar incorrectly.

Reply:
Thanks for the explanation! Changes made.